### PR TITLE
Add retry handling for transient agent failures

### PR DESCRIPTION
**Pull request by mzhaom**
## Summary
- add shared transient error classification for agent providers and Codex API failures
- retry jiradozer agent attempts on transient provider failures, including typed Codex result errors, with configurable `transient_retries` and preserved session IDs for resume
- preserve provider failure details while still returning partial results/session IDs so retry and reporting paths can make correct decisions
- add retry status reporting for backoff, exhaustion, and successful recovery after a retry
- update Codex replay handling to preserve synthetic API errors and recovery status events
- add focused coverage for transient detection, Codex error translation, retry behavior, session resume/preservation, replay behavior, and retry edge cases

## Validation
- `scripts/lint.sh`
- `bazel build //...`
- `bazel test --test_timeout=60 //...`

---

> [!NOTE]
> **Medium Risk**
> Medium risk because it changes execution control flow (retries, session resume, partial results) and error typing across providers and log replay, which could affect reliability and failure reporting.
> 
> **Overview**
> Adds a shared `transient` classifier (`agent-cli-wrapper/transient`) and upgrades Codex turn failures to typed `codex.TransientError` when the error text matches retryable patterns, flowing through `TurnCompletedEvent` and `ParseMappedNotification`.
> 
> Updates `jiradozer` to *automatically retry* transient provider failures (both returned errors and `AgentResult.Error`), with configurable `transient_retries`, backoff messaging, session ID preservation for resume, and per-attempt log-handler state reset to avoid stale plan-file output.
> 
> Adjusts `multiagent` providers to better support retries/resume by returning partial `AgentResult` with `SessionID` on errors (Claude/Codex), adding Codex thread resume, and ensuring replay marks turn-completed failures as errors and derives failed status until a successful retry on the same thread clears it.
> 
> <sup>Reviewed by [Cursor Bugbot](https://cursor.com/bugbot) for commit af201c49a2c11c5166e12d0a9df2ad15cdece6fe. Bugbot is set up for automated code reviews on this repo. Configure [here](https://www.cursor.com/dashboard/bugbot).</sup>

### CODE REVIEW

File: jiradozer/agent.go [hunk 1]
```diff
@@ -382,49 +392,96 @@ func runAgent(ctx context.Context, stepName, prompt string, cfg StepConfig, work
...
+		}
+		timer := time.NewTimer(backoff)
+		select {
+		case <-ctx.Done():
+			if !timer.Stop() {
+				<-timer.C
+			}
+			logHandler.flushText()
+			return StepAgentResult{SessionID: currentResume}, ctx.Err()
+		case <-timer.C:
```
[the remark this answers is not in the document]
Comment (mzhaom) [reply jiradozer/agent.go:486]: Fixed in af201c4.

File: jiradozer/agent.go [hunk 2]
```diff
@@ -382,49 +392,96 @@ func runAgent(ctx context.Context, stepName, prompt string, cfg StepConfig, work
...
 	}
 
-	var opts []agent.ExecuteOption
-	opts = append(opts,
-		agent.WithProviderWorkDir(workDir),
-		agent.WithProviderPermissionMode(cfg.PermissionMode),
-		agent.WithProviderModel(cfg.Model),
-		agent.WithProviderKeepUserSettings(),
-		agent.WithProviderEventHandler(handler),
-	)
```
[the remark this answers is not in the document]
Comment (mzhaom) [reply jiradozer/agent.go:392]: Fixed in af201c4.

File: jiradozer/agent.go [hunk 3]
```diff
@@ -382,49 +392,96 @@ func runAgent(ctx context.Context, stepName, prompt string, cfg StepConfig, work
...
+			select {
+			case <-ctx.Done():
+				if !timer.Stop() {
+					<-timer.C
+				}
+				logHandler.flushText()
+				return StepAgentResult{SessionID: currentResume}, ctx.Err()
+			case <-timer.C:
+			}
+			continue
```
[the remark this answers is not in the document]
Comment (mzhaom) [reply jiradozer/agent.go:455]: Won't fix: low refactor; the two retry branches differ in nil-handling around result.SessionID, parameterising the difference is clearer than collapsing. (Auto-reply from /pr-polish.)

File: multiagent/agent/codex_provider.go [hunk 4]
```diff
@@ -98,21 +104,30 @@ func (p *CodexProvider) Execute(ctx context.Context, prompt string, wtCtx *wt.Wo
...
+		return &AgentResult{SessionID: thread.ID()}, err
 	}
 	select {
 	case <-turnDone:
 	case <-time.After(150 * time.Millisecond):
+		return &AgentResult{SessionID: thread.ID()}, &codex.TransientError{
+			Message:  "turn never reached turn/completed",
+			ThreadID: thread.ID(),
+			TurnID:   thread.CurrentTurnID(),
+		}
```
[the remark this answers is not in the document]
Comment (mzhaom) [reply multiagent/agent/codex_provider.go]: Superseded by af201c4 — the cited code was changed/removed in a later commit. (Auto-reply from /pr-polish.)

File: agent-cli-wrapper/codex/errors.go [hunk 5]
```diff
@@ -106,3 +107,90 @@ func (e *TurnError) Error() string {
...
+		"stream timeout",
+		"stream disconnected",
+		"stream closed",
+	} {
+		if strings.Contains(s, pattern) {
+			return true
+		}
+	}
+	return false
+}
```
[the remark this answers is not in the document]
Comment (mzhaom) [reply agent-cli-wrapper/codex/errors.go]: Superseded by af201c4 — the cited code was changed/removed in a later commit. (Auto-reply from /pr-polish.)

File: jiradozer/agent.go [hunk 6]
```diff
@@ -382,42 +387,52 @@ func runAgent(ctx context.Context, stepName, prompt string, cfg StepConfig, work
...
+			"backoff", backoff,
+		)
+		if renderer != nil {
+			renderer.Status(fmt.Sprintf("Transient error (%s); retrying in %s (attempt %d/%d)", reason, backoff, attempt, maxRetries))
+		}
+		select {
+		case <-ctx.Done():
+			logHandler.flushText()
+			return StepAgentResult{SessionID: currentResume}, ctx.Err()
+		case <-time.After(backoff):
```
[the remark this answers is not in the document]
Comment (mzhaom) [reply jiradozer/agent.go]: Superseded by af201c4 — the cited code was changed/removed in a later commit. (Auto-reply from /pr-polish.)

File: agent-cli-wrapper/transient/transient.go [hunk 7]
```diff
@@ -0,0 +1,49 @@
...
+		return ReasonCodexIncomplete, true
+	case strings.Contains(s, "429"),
+		strings.Contains(s, "rate limit"),
+		strings.Contains(s, "rate limited"):
+		return ReasonHTTP429, true
+	case strings.Contains(s, "500"),
+		strings.Contains(s, "502"),
+		strings.Contains(s, "503"),
+		strings.Contains(s, "504"),
+		strings.Contains(s, "529"):
```
[the remark this answers is not in the document]
Comment (mzhaom) [reply agent-cli-wrapper/transient/transient.go]: Superseded by af201c4 — substring matching for HTTP digits was replaced by a word-boundary regex (httpStatusPattern in agent-cli-wrapper/transient/transient.go) before this comment was filed. (Auto-reply from /pr-polish.)
